perma_web/api2/views.py

```diff
@@ -0,0 +1,483 @@
+
+    def get(self, request, format=None):
+        """ List orgs. """
+        queryset = Organization.objects.accessible_to(request.user).select_related('registrar', 'shared_folder')
+        return self.simple_list(request, queryset)
+
+# /organizations/:id
+class OrganizationDetailView(BaseView):
+    serializer_class = OrganizationSerializer
+
+    def get(self, request, pk, format=None):
+        """ Single org details. """
+        return self.simple_get(request, pk)
+
+
+### FOLDER VIEWS ###
+
+# /folders
+# /folders/:parent_id/folders
+class FolderListView(BaseView):
+    serializer_class = FolderSerializer
+
+    @load_parent
+    def get(self, request, format=None):
+        """ List folders for user. """
+        if request.parent:
+            # for /folders/:parent_id/folders, list subfolders of parent folder
+            queryset = Folder.objects.filter(parent=request.parent)
+        else:
+            # for /folders, list all top level folders for user
+            queryset = request.user.top_level_folders()
+        return self.simple_list(request, queryset)
+
+    @load_parent
+    def post(self, request, format=None):
+        """ Create folder. """
+        # if parent folder is not supplied in post data, try to get it from /folders/:parent_id:
+        data = request.data.copy()
+        if request.parent:
+            data.setdefault('parent', request.parent.pk)
+
+        return self.simple_create(data, {'created_by': request.user})
+
+# /folders/:id
+# /folders/:parent_id/folders/:id
+class FolderDetailView(BaseView):
+    serializer_class = FolderSerializer
+
+    @load_parent
+    def get(self, request, pk, format=None):
+        """ Single folder details. """
+        return self.simple_get(request, pk)
+
+    def folder_update(self, request, pk, data):
+        """ Helper for updating folder details -- used by patch and put methods. """
+        obj = self.get_object_for_user_by_pk(request.user, pk)
+        try:
+            return self.simple_update(obj, data)
+        except InvalidMove as e:
+            raise ValidationError({"parent":[e.args[0]]})
+
+    @load_parent
+    def patch(self, request, pk, format=None):
+        """ Update folder. """
+        return self.folder_update(request, pk, request.data)
+
+    @load_parent
+    def put(self, request, pk, format=None):
+        """
+            Move folder.
+            For nested endpoint, PUT /folders/:id into /folders/:parent_id.
+        """
+        if not request.parent:
+            raise_validation_error("PUT is only valid for nested folder endpoints.")
+        return self.folder_update(request, pk, {'parent': request.parent.pk})
+
+    @load_parent
+    def delete(self, request, pk, format=None):
+        """ Delete folder. """
+        folder = self.get_object_for_user_by_pk(request.user, pk)
+
+        # delete validations
+        if folder.is_shared_folder:
+            raise_validation_error("Shared folders cannot be deleted.")
```

Comment:
Triviality: should we reword this message, since you can delete an org's subfolders if they are empty? How about "Organizations' top-level folders cannot be deleted" or something similar? Or how about combining if folder.is_shared_folder and elif folder.is_root_folder and have them both say "Top-level folders cannot be deleted."